OpenRA.Game/Support/ConditionExpression.cs

```diff
@@ -0,0 +1,898 @@
+				{
+					// Expressions can't start with a binary or unary postfix operation or closer
+					if (token.LeftOperand)
+						throw new InvalidDataException("Missing value or sub-expression at beginning for `{0}` operator".F(token.Symbol));
+				}
+				else
+				{
+					// Disallow empty parentheses
+					if (lastToken.Opens != Grouping.None && token.Closes != Grouping.None)
+						throw new InvalidDataException("Empty parenthesis at index {0}".F(lastToken.Index));
+
+					// Exactly one of two consective tokens must take the other's sub-expression evaluation as an operand
+					if (lastToken.RightOperand == token.LeftOperand)
+					{
+						if (lastToken.RightOperand)
+							throw new InvalidDataException(
+								"Missing value or sub-expression or there is an extra operator `{0}` at index {1} or `{2}` at index {3}".F(
+									lastToken.Symbol, lastToken.Index, token.Symbol, token.Index));
+						throw new InvalidDataException("Missing binary operation before `{0}` at index {1}".F(token.Symbol, token.Index));
+					}
+				}
+
+				if (token.Type == TokenType.Variable)
+					variables.Add(token.Symbol);
+
+				tokens.Add(token);
+				lastToken = token;
+			}
+
+			if (currentOpeners.Count > 0)
+				throw new InvalidDataException("Unclosed opening parenthesis at index {0}".F(currentOpeners.Peek().Index));
+
+			asFunction = new Compiler().Compile(ToPostfix(tokens).ToArray());
+		}
+
+		static int ParseSymbol(string symbol, IReadOnlyDictionary<string, int> symbols)
+		{
+			int value;
+			symbols.TryGetValue(symbol, out value);
+			return value;
+		}
+
+		static IEnumerable<Token> ToPostfix(IEnumerable<Token> tokens)
+		{
+			var s = new Stack<Token>();
+			foreach (var t in tokens)
+			{
+				if (t.Opens != Grouping.None)
+					s.Push(t);
+				else if (t.Closes != Grouping.None)
+				{
+					Token temp;
+					while (!((temp = s.Pop()).Opens != Grouping.None))
+						yield return temp;
+				}
+				else if (t.OperandSides == OperandSides.None)
+					yield return t;
+				else
+				{
+					while (s.Count > 0 && ((t.Associativity == Associativity.Right && t.Precedence < s.Peek().Precedence)
+						|| (t.Associativity == Associativity.Left && t.Precedence <= s.Peek().Precedence)))
+						yield return s.Pop();
+
+					s.Push(t);
+				}
+			}
+
+			while (s.Count > 0)
+				yield return s.Pop();
+		}
+
+		enum ExpressionType { Int, Bool }
+
+		static readonly ParameterExpression SymbolsParam =
+			Expressions.Expression.Parameter(typeof(IReadOnlyDictionary<string, int>), "symbols");
+		static readonly ConstantExpression Zero = Expressions.Expression.Constant(0);
+		static readonly ConstantExpression One = Expressions.Expression.Constant(1);
+		static readonly ConstantExpression False = Expressions.Expression.Constant(false);
+		static readonly ConstantExpression True = Expressions.Expression.Constant(true);
+
+		static Expression AsBool(Expression expression)
+		{
+			return Expressions.Expression.GreaterThan(expression, Zero);
```

Comment:
I'd assume !condition would mean the expression is valid when the condition level is 0 and any other cases I would expect a relation be !d, so yes, I agree.